torch/csrc/distributed/rpc/python_functions.cpp

```diff
@@ -74,13 +77,20 @@ std::shared_ptr<Operator> matchBuiltinOp(
         continue;
       }
 
-      // Found the right op!
-      return op;
+      // Prefer C10 ops so that they go through C10 dispatch. We expect the
+      // total # of possible overloaded ops to be small (i.e. it is 10 for
+      // torch.add) so a worst-case linear search should not incur significant
+      // extra overhead.
+      if (op->isC10Op()) {
+        return op;
```

Comment:
Thanks for fixing! And thanks for digging into the bottom of this error!

This op-matching logic does need a thorough revamp.